File: hpg.py
#!/usr/bin/env python
from socket import socket, AF_UNIX
import sys
import os
import os.path

from bottle import route, run, static_file, request


SOCK_ADDR = "/home/henry/.config/mpd/socket"
SERVER_ROOT = "/home/henry/Entwicklung/HyProGlo-purescript/server"


def _mpd(query, sock_addr=SOCK_ADDR):
    result = None

    sock = socket(AF_UNIX)
    sock.connect(sock_addr)
    status = sock.recv(64)
    if status.startswith(b"OK MPD"):
        sock.sendall(query)
        result = sock.recv(4096)
        if not result.startswith(b"ACK"):
            while not result.endswith(b"OK\n"):
                result += sock.recv(4096)
    sock.close()
    return status, result


@route('/')
def home():
    return static_file('index.html', root=SERVER_ROOT)


@route('/mpd', method='POST')
def mpd():
    answer = {"status": None, "result": None, "error": None}
    try:
        query = request.body.read()
        if query[-1:] != b"\n":
            query += b"\n"
        status, result = _mpd(query)
        answer["status"] = status.decode()
        answer["result"] = result.decode()
    except:
        answer["error"] = str(sys.exc_info()[1])
        raise
    return answer


MPD_ROOT = None
IMAGES = [".jpg", ".jpeg", ".png", ".gif"]

@route('/image/<path:path>')
def image(path):
    global MPD_ROOT
    if MPD_ROOT is None:
        _, result = _mpd(b"config\n")
        for line in result.decode().splitlines():
            if line.startswith("music_directory: "):
                MPD_ROOT = line.split(": ", 1)[1]

    path = os.path.dirname(os.path.join(MPD_ROOT, path))
    images = [f for f in os.listdir(path) if os.path.splitext(f)[1].lower() in IMAGES]
    if len(images) == 0:
        return None
    else:
        full = os.path.join(path, images[0])
        relpath = os.path.relpath(full, start=MPD_ROOT)
        print(MPD_ROOT, full, relpath)
        return static_file(relpath, root=MPD_ROOT)


@route('/<path:path>')
def static(path):
    return static_file(path, root=SERVER_ROOT)


run(host='0.0.0.0', port=5000, server='cherrypy')

File: hpg.go
package main
import (
    "encoding/json"
    "fmt"
    "io/ioutil"
    "log"
    "net"
    "net/http"
    "os"
    "path"
    "strings"
)


var socketAddress string
var mpdRoot *string


type mpdReply struct {
    Status string `json:"status"`
    Result string `json:"result"`
    Error error `json:"error"`
}


func mpd(q string, addr string) (string, error) {
    if !strings.HasSuffix(q, "\n") {
        q += "\n"
    }

    conn, err := net.Dial("unix", addr)
    if err != nil {
        return "", err
    }

    var buf = make([]byte, 4 * 1024)
    read, err := conn.Read(buf)
    if read < 1 || err != nil {
        conn.Close()
        return "", err
    }

    status := string(buf[:read])
    if !strings.HasPrefix(status, "OK MPD") {
        conn.Close()
        return "", fmt.Errorf("Unexcpected MPD status: %s", status)
    }

    conn.Write([]byte(q))
    read, err = conn.Read(buf)
    if err != nil {
        return "", err
    }

    reply := string(buf[:read])
    if strings.HasPrefix(reply, "ACK") {
        return "", fmt.Errorf("MPD error: %s", reply)
    }

    for !strings.HasSuffix(reply, "OK\n") {
        read, err = conn.Read(buf)
        if err != nil {
            return "", err
        }
        reply += string(buf[:read])
    }
    conn.Close()
    return reply, nil
}


func mpdHandler(w http.ResponseWriter, r *http.Request) {
    if r.Method != "POST" {
        http.Error(w, "Unsupported method, use POST.", 405)
        return
    }

    data, err := ioutil.ReadAll(r.Body)
    if err != nil {
        http.Error(w, err.Error(), 400)
        return
    }

    userQuery := string(data)
    reply, err := mpd(userQuery, socketAddress)
    js, err := json.Marshal(mpdReply{"", reply, err})
    if err != nil {
        http.Error(w, err.Error(), 500)
        return
    }
    fmt.Fprint(w, string(js))
}


func imageHandler(w http.ResponseWriter, r *http.Request) {
    if mpdRoot == nil {
        res, err := mpd("config", socketAddress)
        if err != nil {
            http.Error(w, err.Error(), 503)
            return
        }
        root := strings.TrimPrefix(res, "music_directory: ")
        root = strings.TrimSuffix(root, "\nOK\n")
        mpdRoot = &root
    }
    img_path := strings.TrimPrefix(r.URL.Path, "/image/")
    img_path = path.Dir(path.Join(*mpdRoot, img_path))

    files, err := ioutil.ReadDir(img_path)
    if err != nil {
        http.Error(w, err.Error(), 404)
        return
    }

    for _, file := range files {
        img_file := file.Name()
        ext := strings.ToLower(path.Ext(img_file))
        if ext == ".jpg" || ext == ".jpeg" || ext == ".png" || ext == ".gif" {
            http.ServeFile(w, r, path.Join(img_path, img_file))
            return
        }
    }
}


func main() {
    if len(os.Args) != 3 {
        fmt.Printf("Usage: %s PORT MPD_SOCKET\n", os.Args[0])
        os.Exit(1)
    }
    port := os.Args[1]
    socketAddress = os.Args[2]

    fs := http.FileServer(http.Dir("server"))
    http.Handle("/", fs)
    http.HandleFunc("/mpd", mpdHandler)
    http.HandleFunc("/image/", imageHandler)

    log.Fatal(http.ListenAndServe(":" + port, nil))
}
